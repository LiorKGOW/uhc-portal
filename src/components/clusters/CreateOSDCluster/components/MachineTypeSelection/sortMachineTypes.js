// This is a helper function to sort a MachineType structure according to size and category.
import { parseValueWithUnit } from '../../../../../common/units';

function sortByMemory(a, b) {
  // sort by memory, ascending order
  const memoryBytesA = parseValueWithUnit(a.memory.value, a.memory.unit);
  const memoryBytesB = parseValueWithUnit(b.memory.value, b.memory.unit);
  return memoryBytesA - memoryBytesB;
}

function sortMachineTypes(a, b) {
  // Sort by category, so the resulting order is m (general), r (memory), c (compute)
  const categoryA = a.id.toLowerCase().charAt(0);
  const categoryB = b.id.toLowerCase().charAt(0);
  if (categoryA === categoryB) {
    // the categories are equal. Within the category, we need to sort by size.
    const memorySortResult = sortByMemory(a, b);
    if (memorySortResult === 0) {
      // sort by CPU if two machine types have the same amount of memory
      return a.cpu.value - b.cpu.value;
    }
    return memorySortResult;
  }
  if (categoryA === 'm') {
    return -1;
  }
  if (categoryB === 'm') {
    return 1;
  }
  if (categoryA === 'r' && categoryB === 'c') {
    return -1;
  }
  if (categoryA === 'c' && categoryB === 'r') {
    return 1;
  }
  return 0;
}


export default sortMachineTypes;
