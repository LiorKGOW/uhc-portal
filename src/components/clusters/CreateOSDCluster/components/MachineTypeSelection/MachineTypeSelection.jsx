// MachineTypeSelector renders a series of radio buttons for all available node types,
// allowing the user to select just one.
// It is meant to be used in a redux-form <Field> and expects an onChange callback.

import React from 'react';
import PropTypes from 'prop-types';
import { Gallery, GalleryItem } from '@patternfly/react-core';
import { CpuIcon, MemoryIcon, ContainerNodeIcon } from '@patternfly/react-icons';
import { Spinner } from '@redhat-cloud-services/frontend-components';
import FlatRadioButton from '../../../../common/FlatRadioButton';
import ErrorBox from '../../../../common/ErrorBox';
import { humanizeValueWithUnit } from '../../../../../common/units';
import sortMachineTypes from './sortMachineTypes';

const machineTypeIcon = (machineType) => {
  const machineTypeFamily = machineType.charAt(0);
  switch (machineTypeFamily) {
    case 'r':
      return <MemoryIcon size="lg" />;
    case 'c':
      return <CpuIcon size="lg" />;
    case 'm':
      return <ContainerNodeIcon size="lg" />;
    default:
      return null;
  }
};

class MachineTypeSelection extends React.Component {
  state = {
    currentValue: '',
  }

  componentDidMount() {
    const { getMachineTypes, machineTypes } = this.props;
    if (!machineTypes.fulfilled) {
      // Don't let the user submit if we couldn't get machine types yet.
      this.setInvalidValue();
    }
    if (!machineTypes.pending && !machineTypes.fulfilled) {
      // fetch cloud providers from server only if needed.
      getMachineTypes();
    }

    if (machineTypes.fulfilled) {
      this.setDefaultValue();
    }
  }

  componentDidUpdate() {
    const { machineTypes } = this.props;
    const { currentValue } = this.state;
    if (machineTypes.error || machineTypes.pending || !this.hasQuota(currentValue)) {
      // Don't let the user submit if we couldn't get machine types.
      this.setInvalidValue();
    }

    if (currentValue === '' && machineTypes.fulfilled) {
      // we got the machine types, and the user hasn't selected one yet - set to default.
      this.setDefaultValue();
    }

    // if some external param changed, like MultiAz, and we no longer have quota
    // for the selected instance type, we need to unselect it, and mark ourselves as invalid.
    if (currentValue && !this.hasQuota(currentValue)) {
      // this setState is guarded so the linter error can be ignored.
      // eslint-disable-next-line react/no-did-update-set-state
      this.setState({ currentValue: '' });
    }
  }

  setDefaultValue() {
    // Find the first machineType we have quota for, and set it as default
    const { machineTypes, input } = this.props;
    const defaultType = machineTypes.types.find(type => this.hasQuota(type.id));
    if (defaultType) {
      this.setState({ currentValue: defaultType.id });
      input.onChange(defaultType.id);
    }
  }

  setInvalidValue() {
    // Tell redux form the current value of this field is empty.
    // This will cause it to not pass validation if it is required.
    const { input } = this.props;
    input.onChange('');
  }

  hasQuota(machineType) {
    const { isMultiAz, quota } = this.props;
    if (!quota.fulfilled) {
      return false;
    }
    const available = quota.quotaList.nodeQuota.rhInfra[isMultiAz ? 'multiAz' : 'singleAz'][machineType] || 0;
    return available > 0;
  }

  render() {
    // getMachineTypes is unused here, but it's needed so it won't
    // go into extraProps and then get to the DOM, generating a React warning.
    const {
      machineTypes,
      getMachineTypes,
      isMultiAz,
      quota,
      input,
      meta: { error, touched },
      ...extraProps
    } = this.props;
    const { currentValue } = this.state;


    const changeHandler = (value) => {
      this.setState({ currentValue: value });
      input.onChange(value);
    };

    const machineTypeRadio = (machineType) => {
      const humanizedMemory = humanizeValueWithUnit(machineType.memory.value,
        machineType.memory.unit);
      const labelTitle = `${machineType.cpu.value} ${machineType.cpu.unit} ${humanizedMemory.value} ${humanizedMemory.unit} RAM`;

      const hasQuota = this.hasQuota(machineType.id);
      return (
        <GalleryItem key={machineType.id}>
          <FlatRadioButton
            {...extraProps}
            id={`machineTypeRadio.${machineType.id}`}
            value={machineType.id}
            isDisabled={!hasQuota}
            isSelected={hasQuota && currentValue === machineType.id}
            titleText={labelTitle}
            secondaryText={machineType.name}
            icon={machineTypeIcon(machineType.id)}
            onChange={changeHandler}
          />
        </GalleryItem>
      );
    };

    if (machineTypes.fulfilled) {
      machineTypes.types.sort(sortMachineTypes);
      return (
        <div className="node-type-input">
          <div className="node-type-label">Node type</div>
          {(touched && error) && (<span className="error">{error}</span>)}
          <Gallery gutter="sm">
            {machineTypes.types.map(type => machineTypeRadio(type))}
          </Gallery>
        </div>
      );
    }

    return machineTypes.error ? (
      <ErrorBox message="Error loading node types" response={machineTypes} />
    ) : (
      <React.Fragment>
        <div className="spinner-fit-container"><Spinner /></div>
        <div className="spinner-loading-text">Loading node types..</div>
      </React.Fragment>
    );
  }
}

MachineTypeSelection.propTypes = {
  input: PropTypes.shape({ onChange: PropTypes.func.isRequired }).isRequired,
  getMachineTypes: PropTypes.func.isRequired,
  machineTypes: PropTypes.object.isRequired,
  isMultiAz: PropTypes.bool.isRequired,
  quota: PropTypes.object.isRequired,
  // Plus extraprops passed by redux Field
};

export default MachineTypeSelection;
