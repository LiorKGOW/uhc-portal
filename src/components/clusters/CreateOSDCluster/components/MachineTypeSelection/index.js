import { connect } from 'react-redux';
import MachineTypeSelection from './MachineTypeSelection';
import { getMachineTypes } from '../../../../../redux/actions/machineTypesActions';
import { getOrganizationAndQuota } from '../../../../../redux/actions/userActions';

const mapStateToProps = state => ({
  machineTypes: state.machineTypes,
  organization: state.userProfile.organization,
  quota: state.userProfile.organization.quotaList,
});

const mapDispatchToProps = {
  getOrganizationAndQuota,
  getMachineTypes,
};

export default connect(mapStateToProps, mapDispatchToProps)(MachineTypeSelection);
