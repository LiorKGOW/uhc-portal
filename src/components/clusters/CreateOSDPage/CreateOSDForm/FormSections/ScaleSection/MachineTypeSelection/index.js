import { connect } from 'react-redux';
import MachineTypeSelection from './MachineTypeSelection';
import { getMachineTypes } from '../../../../../../../redux/actions/machineTypesActions';
import { getDefaultFlavour } from '../../../../../../../redux/actions/flavourActions';

const mapStateToProps = state => ({
  flavours: state.flavours,
  machineTypes: state.machineTypes,
  organization: state.userProfile.organization,
  quota: state.userProfile.organization.quotaList,
});

const mapDispatchToProps = {
  getMachineTypes,
  getDefaultFlavour,
};

export default connect(mapStateToProps, mapDispatchToProps)(MachineTypeSelection);
