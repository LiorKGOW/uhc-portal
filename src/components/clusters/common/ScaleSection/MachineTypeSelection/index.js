import { connect } from 'react-redux';
import { getMachineTypes } from '~/redux/actions/machineTypesActions';
import { getDefaultFlavour } from '~/redux/actions/flavourActions';
import MachineTypeSelection from './MachineTypeSelection';

const mapStateToProps = (state) => ({
  flavours: state.flavours,
  machineTypes: state.machineTypes,
  organization: state.userProfile.organization,
  quota: state.userProfile.organization.quotaList,
});

const mapDispatchToProps = {
  getMachineTypes,
  getDefaultFlavour,
};

export default connect(mapStateToProps, mapDispatchToProps)(MachineTypeSelection);
