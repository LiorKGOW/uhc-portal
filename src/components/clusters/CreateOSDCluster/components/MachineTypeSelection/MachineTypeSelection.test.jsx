import React from 'react';
import { mount } from 'enzyme';

import MachineTypeSelection from './MachineTypeSelection';

const baseState = {
  error: false,
  errorMessage: '',
  pending: false,
  fulfilled: false,
  machineTypes: {},
};

const organizationState = {
  fulfilled: true,
  pending: false,
};

describe('<MachineTypeSelection />', () => {
  describe('when machine type list needs to be fetched', () => {
    let getMachineTypes;
    let onChange;
    let wrapper;
    beforeAll(() => {
      getMachineTypes = jest.fn();
      onChange = jest.fn();
      wrapper = mount(
        <MachineTypeSelection
          machineTypes={baseState}
          input={{ onChange }}
          meta={{}}
          getMachineTypes={getMachineTypes}
          isMultiAz={false}
          quota={{}}
          organization={organizationState}
        />,
      );
    });

    it('renders correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('calls getCloudProviders', () => {
      expect(getMachineTypes).toBeCalled();
    });

    it('calls onChange to mark as invalid', () => {
      expect(onChange).toBeCalledWith('');
    });
  });

  describe('when there was an error', () => {
    let getMachineTypes;
    let onChange;
    let wrapper;
    beforeAll(() => {
      const state = {
        ...baseState,
        error: true,
        errorMessage: 'This is an error message',
      };

      getMachineTypes = jest.fn();
      onChange = jest.fn();
      wrapper = mount(
        <MachineTypeSelection
          machineTypes={state}
          input={{ onChange }}
          meta={{}}
          getMachineTypes={getMachineTypes}
          isMultiAz={false}
          quota={{}}
          organization={organizationState}
        />,
      );
    });

    it('renders correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('calls getMachineTypes on mount', () => {
      expect(getMachineTypes).toBeCalled();
    });

    it('calls onChange to mark as invalid', () => {
      expect(onChange).toBeCalledWith('');
    });
  });

  describe('when the request is pending', () => {
    let getMachineTypes;
    let onChange;
    let wrapper;
    const state = {
      error: false,
      errorMessage: '',
      pending: true,
      fulfilled: false,
      types: [],
    };
    beforeAll(() => {
      getMachineTypes = jest.fn();
      onChange = jest.fn();
      wrapper = mount(
        <MachineTypeSelection
          machineTypes={state}
          input={{ onChange }}
          meta={{}}
          getMachineTypes={getMachineTypes}
          isMultiAz={false}
          quota={{}}
          organization={organizationState}
        />,
      );
    });

    it('renders correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('does not call getCloudProviders', () => {
      expect(getMachineTypes).not.toBeCalled();
    });

    it('calls onChange to mark as invalid', () => {
      expect(onChange).toBeCalledWith('');
    });

    it('does not call getMachineTypes again if request returns an error', () => {
      wrapper.setProps({
        machineTypes: {
          ...state,
          error: true,
          pending: false,
        },
      }, () => {
        expect(getMachineTypes).not.toBeCalled();
      });
    });
  });

  describe('when the machine types list is available', () => {
    let getMachineTypes;
    let onChange;
    let wrapper;
    beforeAll(() => {
      const state = {
        ...baseState,
        fulfilled: true,
      };

      const sortedMachineTypes = [
        {
          kind: 'MachineType',
          name: 'Memory optimized - R5.XLarge',
          id: 'r5.xlarge',
          href: '/api/clusters_mgmt/v1/machine_types/r5.xlarge',
          memory: {
            value: 34359738368,
            unit: 'B',
          },
          cpu: {
            value: 4,
            unit: 'vCPU',
          },
          cloud_provider: {
            kind: 'CloudProviderLink',
            id: 'aws',
            href: '/api/clusters_mgmt/v1/cloud_providers/aws',
          },
        },
        {
          kind: 'MachineType',
          name: 'Memory optimized - R5.4XLarge',
          id: 'r5.4xlarge',
          href: '/api/clusters_mgmt/v1/machine_types/r5.4xlarge',
          memory: {
            value: 137438953472,
            unit: 'B',
          },
          cpu: {
            value: 16,
            unit: 'vCPU',
          },
          cloud_provider: {
            kind: 'CloudProviderLink',
            id: 'aws',
            href: '/api/clusters_mgmt/v1/cloud_providers/aws',
          },
        },
      ];

      const quota = {
        clusterQuota: {
          rhInfra: {
            multiAz: {
              'r5.xlarge': 5,
            },
            singleAz: {
              'r5.xlarge': 0,
            },
          },
        },
      };

      getMachineTypes = jest.fn();
      onChange = jest.fn();
      wrapper = mount(
        <MachineTypeSelection
          machineTypes={state}
          sortedMachineTypes={sortedMachineTypes}
          input={{ onChange }}
          meta={{}}
          getMachineTypes={getMachineTypes}
          quota={quota}
          organization={organizationState}
          isMultiAz
        />,
      );
    });

    it('renders correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('does not call getMachineTypes', () => {
      expect(getMachineTypes).not.toBeCalled();
    });

    it('calls onChange with the first item that has quota', () => {
      expect(onChange).toBeCalledWith('r5.xlarge');
    });
  });
});
