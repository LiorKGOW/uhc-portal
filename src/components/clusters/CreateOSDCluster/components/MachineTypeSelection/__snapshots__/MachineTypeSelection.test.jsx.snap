// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MachineTypeSelection /> when machine type list needs to be fetched renders correctly 1`] = `
<MachineTypeSelection
  getMachineTypes={
    [MockFunction] {
      "calls": Array [
        Array [],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  input={
    Object {
      "onChange": [MockFunction] {
        "calls": Array [
          Array [
            "",
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      },
    }
  }
  isMultiAz={false}
  machineTypes={
    Object {
      "error": false,
      "errorMessage": "",
      "fulfilled": false,
      "machineTypes": Array [],
      "pending": false,
    }
  }
  meta={Object {}}
  organization={
    Object {
      "fulfilled": true,
      "pending": false,
    }
  }
  quota={Object {}}
>
  <div
    className="spinner-fit-container"
  >
    <Spinner>
      <div
        className="ins-c-spinner"
        role="status"
      >
        <span
          className="pf-u-screen-reader"
        >
          Loading...
        </span>
      </div>
    </Spinner>
  </div>
  <div
    className="spinner-loading-text"
  >
    Loading node types...
  </div>
</MachineTypeSelection>
`;

exports[`<MachineTypeSelection /> when the machine types list is available renders correctly 1`] = `
<MachineTypeSelection
  getMachineTypes={[MockFunction]}
  input={
    Object {
      "onChange": [MockFunction] {
        "calls": Array [
          Array [
            "r5.xlarge",
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      },
    }
  }
  isMultiAz={true}
  machineTypes={
    Object {
      "error": false,
      "errorMessage": "",
      "fulfilled": true,
      "machineTypes": Array [],
      "pending": false,
      "types": Array [
        Object {
          "cloud_provider": Object {
            "href": "/api/clusters_mgmt/v1/cloud_providers/aws",
            "id": "aws",
            "kind": "CloudProviderLink",
          },
          "cpu": Object {
            "unit": "vCPU",
            "value": 4,
          },
          "href": "/api/clusters_mgmt/v1/machine_types/r5.xlarge",
          "id": "r5.xlarge",
          "kind": "MachineType",
          "memory": Object {
            "unit": "B",
            "value": 34359738368,
          },
          "name": "Memory optimized - R5.XLarge",
        },
        Object {
          "cloud_provider": Object {
            "href": "/api/clusters_mgmt/v1/cloud_providers/aws",
            "id": "aws",
            "kind": "CloudProviderLink",
          },
          "cpu": Object {
            "unit": "vCPU",
            "value": 16,
          },
          "href": "/api/clusters_mgmt/v1/machine_types/r5.4xlarge",
          "id": "r5.4xlarge",
          "kind": "MachineType",
          "memory": Object {
            "unit": "B",
            "value": 137438953472,
          },
          "name": "Memory optimized - R5.4XLarge",
        },
      ],
    }
  }
  meta={Object {}}
  organization={
    Object {
      "fulfilled": true,
      "pending": false,
    }
  }
  quota={
    Object {
      "clusterQuota": Object {
        "rhInfra": Object {
          "multiAz": Object {
            "r5.xlarge": 5,
          },
          "singleAz": Object {
            "r5.xlarge": 0,
          },
        },
      },
    }
  }
>
  <div
    className="flat-radio-buttons-flex-container"
  >
    <FlatRadioButton
      icon={
        <MemoryIcon
          color="currentColor"
          noVerticalAlign={false}
          size="lg"
          title={null}
        />
      }
      id="machineTypeRadio.r5.xlarge"
      isDisabled={false}
      isSelected={true}
      key="r5.xlarge"
      onChange={[Function]}
      secondaryText="Memory optimized - R5.XLarge"
      titleText="4 vCPU 32 GiB RAM"
      tooltip={false}
      value="r5.xlarge"
    >
      <Component
        className="ocm-flat-radio-button selected"
        id="machineTypeRadio.r5.xlarge"
        isDisabled={false}
        onClick={[Function]}
        variant="tertiary"
      >
        <ComponentWithOuia
          component={[Function]}
          componentProps={
            Object {
              "children": <Split
                gutter="sm"
              >
                <SplitItem
                  className="ocm-flat-button-icon"
                >
                  <MemoryIcon
                    color="currentColor"
                    noVerticalAlign={false}
                    size="lg"
                    title={null}
                  />
                </SplitItem>
                <SplitItem
                  isFilled={true}
                >
                  <Title
                    headingLevel="h4"
                    size="lg"
                  >
                    4 vCPU 32 GiB RAM
                  </Title>
                  Memory optimized - R5.XLarge
                </SplitItem>
              </Split>,
              "className": "ocm-flat-radio-button selected",
              "id": "machineTypeRadio.r5.xlarge",
              "isDisabled": false,
              "onClick": [Function],
              "variant": "tertiary",
            }
          }
          consumerContext={null}
        >
          <Button
            className="ocm-flat-radio-button selected"
            id="machineTypeRadio.r5.xlarge"
            isDisabled={false}
            onClick={[Function]}
            ouiaContext={
              Object {
                "isOuia": false,
                "ouiaId": null,
              }
            }
            variant="tertiary"
          >
            <button
              aria-disabled={null}
              aria-label={null}
              className="pf-c-button pf-m-tertiary ocm-flat-radio-button selected"
              disabled={false}
              id="machineTypeRadio.r5.xlarge"
              onClick={[Function]}
              tabIndex={null}
              type="button"
            >
              <Split
                gutter="sm"
              >
                <div
                  className="pf-l-split pf-m-gutter"
                >
                  <SplitItem
                    className="ocm-flat-button-icon"
                  >
                    <div
                      className="pf-l-split__item ocm-flat-button-icon"
                    >
                      <MemoryIcon
                        color="currentColor"
                        noVerticalAlign={false}
                        size="lg"
                        title={null}
                      >
                        <svg
                          aria-hidden={true}
                          aria-labelledby={null}
                          fill="currentColor"
                          height="2em"
                          role="img"
                          style={
                            Object {
                              "verticalAlign": "-0.25em",
                            }
                          }
                          viewBox="0 0 640 512"
                          width="2em"
                        >
                          <path
                            d="M640 130.94V96c0-17.67-14.33-32-32-32H32C14.33 64 0 78.33 0 96v34.94c18.6 6.61 32 24.19 32 45.06s-13.4 38.45-32 45.06V320h640v-98.94c-18.6-6.61-32-24.19-32-45.06s13.4-38.45 32-45.06zM224 256h-64V128h64v128zm128 0h-64V128h64v128zm128 0h-64V128h64v128zM0 448h64v-26.67c0-8.84 7.16-16 16-16s16 7.16 16 16V448h128v-26.67c0-8.84 7.16-16 16-16s16 7.16 16 16V448h128v-26.67c0-8.84 7.16-16 16-16s16 7.16 16 16V448h128v-26.67c0-8.84 7.16-16 16-16s16 7.16 16 16V448h64v-96H0v96z"
                            transform=""
                          />
                        </svg>
                      </MemoryIcon>
                    </div>
                  </SplitItem>
                  <SplitItem
                    isFilled={true}
                  >
                    <div
                      className="pf-l-split__item pf-m-fill"
                    >
                      <Title
                        headingLevel="h4"
                        size="lg"
                      >
                        <h4
                          className="pf-c-title pf-m-lg"
                        >
                          4 vCPU 32 GiB RAM
                        </h4>
                      </Title>
                      Memory optimized - R5.XLarge
                    </div>
                  </SplitItem>
                </div>
              </Split>
            </button>
          </Button>
        </ComponentWithOuia>
      </Component>
    </FlatRadioButton>
    <FlatRadioButton
      icon={
        <MemoryIcon
          color="currentColor"
          noVerticalAlign={false}
          size="lg"
          title={null}
        />
      }
      id="machineTypeRadio.r5.4xlarge"
      isDisabled={true}
      isSelected={false}
      key="r5.4xlarge"
      onChange={[Function]}
      secondaryText="Memory optimized - R5.4XLarge"
      titleText="16 vCPU 128 GiB RAM"
      tooltip="You do not have quota for this node type. Contact sales to purchase additional quota."
      value="r5.4xlarge"
    >
      <Tooltip
        appendTo={[Function]}
        aria="describedby"
        boundary="window"
        className=""
        content="You do not have quota for this node type. Contact sales to purchase additional quota."
        distance={15}
        enableFlip={true}
        entryDelay={500}
        exitDelay={500}
        flipBehavior={
          Array [
            "top",
            "right",
            "bottom",
            "left",
            "top",
            "right",
            "bottom",
          ]
        }
        id=""
        isAppLauncher={false}
        isContentLeftAligned={false}
        isVisible={false}
        maxWidth="18.75rem"
        position="top"
        tippyProps={Object {}}
        trigger="mouseenter focus"
        zIndex={9999}
      >
        <PopoverBase
          appendTo={[Function]}
          aria="describedby"
          arrow={true}
          boundary="window"
          content={
            <div
              className=""
              id=""
              role="tooltip"
            >
              <TooltipContent
                isLeftAligned={false}
              >
                You do not have quota for this node type. Contact sales to purchase additional quota.
              </TooltipContent>
            </div>
          }
          delay={
            Array [
              500,
              500,
            ]
          }
          distance={15}
          flip={true}
          flipBehavior={
            Array [
              "top",
              "right",
              "bottom",
              "left",
              "top",
              "right",
              "bottom",
            ]
          }
          isVisible={false}
          lazy={true}
          maxWidth="18.75rem"
          onCreate={[Function]}
          placement="top"
          popperOptions={
            Object {
              "modifiers": Object {
                "hide": Object {
                  "enabled": true,
                },
                "preventOverflow": Object {
                  "enabled": true,
                },
              },
            }
          }
          theme="pf-tooltip"
          trigger="mouseenter focus"
          zIndex={9999}
        >
          <div
            className="ocm-flat-radio-button-disabled"
          >
            <Component
              className="ocm-flat-radio-button"
              id="machineTypeRadio.r5.4xlarge"
              isDisabled={true}
              onClick={[Function]}
              variant="tertiary"
            >
              <ComponentWithOuia
                component={[Function]}
                componentProps={
                  Object {
                    "children": <Split
                      gutter="sm"
                    >
                      <SplitItem
                        className="ocm-flat-button-icon"
                      >
                        <MemoryIcon
                          color="currentColor"
                          noVerticalAlign={false}
                          size="lg"
                          title={null}
                        />
                      </SplitItem>
                      <SplitItem
                        isFilled={true}
                      >
                        <Title
                          headingLevel="h4"
                          size="lg"
                        >
                          16 vCPU 128 GiB RAM
                        </Title>
                        Memory optimized - R5.4XLarge
                      </SplitItem>
                    </Split>,
                    "className": "ocm-flat-radio-button",
                    "id": "machineTypeRadio.r5.4xlarge",
                    "isDisabled": true,
                    "onClick": [Function],
                    "variant": "tertiary",
                  }
                }
                consumerContext={null}
              >
                <Button
                  className="ocm-flat-radio-button"
                  id="machineTypeRadio.r5.4xlarge"
                  isDisabled={true}
                  onClick={[Function]}
                  ouiaContext={
                    Object {
                      "isOuia": false,
                      "ouiaId": null,
                    }
                  }
                  variant="tertiary"
                >
                  <button
                    aria-disabled={null}
                    aria-label={null}
                    className="pf-c-button pf-m-tertiary ocm-flat-radio-button"
                    disabled={true}
                    id="machineTypeRadio.r5.4xlarge"
                    onClick={[Function]}
                    tabIndex={null}
                    type="button"
                  >
                    <Split
                      gutter="sm"
                    >
                      <div
                        className="pf-l-split pf-m-gutter"
                      >
                        <SplitItem
                          className="ocm-flat-button-icon"
                        >
                          <div
                            className="pf-l-split__item ocm-flat-button-icon"
                          >
                            <MemoryIcon
                              color="currentColor"
                              noVerticalAlign={false}
                              size="lg"
                              title={null}
                            >
                              <svg
                                aria-hidden={true}
                                aria-labelledby={null}
                                fill="currentColor"
                                height="2em"
                                role="img"
                                style={
                                  Object {
                                    "verticalAlign": "-0.25em",
                                  }
                                }
                                viewBox="0 0 640 512"
                                width="2em"
                              >
                                <path
                                  d="M640 130.94V96c0-17.67-14.33-32-32-32H32C14.33 64 0 78.33 0 96v34.94c18.6 6.61 32 24.19 32 45.06s-13.4 38.45-32 45.06V320h640v-98.94c-18.6-6.61-32-24.19-32-45.06s13.4-38.45 32-45.06zM224 256h-64V128h64v128zm128 0h-64V128h64v128zm128 0h-64V128h64v128zM0 448h64v-26.67c0-8.84 7.16-16 16-16s16 7.16 16 16V448h128v-26.67c0-8.84 7.16-16 16-16s16 7.16 16 16V448h128v-26.67c0-8.84 7.16-16 16-16s16 7.16 16 16V448h128v-26.67c0-8.84 7.16-16 16-16s16 7.16 16 16V448h64v-96H0v96z"
                                  transform=""
                                />
                              </svg>
                            </MemoryIcon>
                          </div>
                        </SplitItem>
                        <SplitItem
                          isFilled={true}
                        >
                          <div
                            className="pf-l-split__item pf-m-fill"
                          >
                            <Title
                              headingLevel="h4"
                              size="lg"
                            >
                              <h4
                                className="pf-c-title pf-m-lg"
                              >
                                16 vCPU 128 GiB RAM
                              </h4>
                            </Title>
                            Memory optimized - R5.4XLarge
                          </div>
                        </SplitItem>
                      </div>
                    </Split>
                  </button>
                </Button>
              </ComponentWithOuia>
            </Component>
          </div>
          <Portal
            containerInfo={
              <div>
                <div
                  class=""
                  id=""
                  role="tooltip"
                >
                  <div
                    class="pf-c-tooltip__content"
                  >
                    You do not have quota for this node type. Contact sales to purchase additional quota.
                  </div>
                </div>
              </div>
            }
          >
            <div
              className=""
              id=""
              role="tooltip"
            >
              <TooltipContent
                isLeftAligned={false}
              >
                <div
                  className="pf-c-tooltip__content"
                >
                  You do not have quota for this node type. Contact sales to purchase additional quota.
                </div>
              </TooltipContent>
            </div>
          </Portal>
        </PopoverBase>
      </Tooltip>
    </FlatRadioButton>
  </div>
</MachineTypeSelection>
`;

exports[`<MachineTypeSelection /> when the request is pending renders correctly 1`] = `
<MachineTypeSelection
  getMachineTypes={[MockFunction]}
  input={
    Object {
      "onChange": [MockFunction] {
        "calls": Array [
          Array [
            "",
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      },
    }
  }
  isMultiAz={false}
  machineTypes={
    Object {
      "error": false,
      "errorMessage": "",
      "fulfilled": false,
      "pending": true,
      "types": Array [],
    }
  }
  meta={Object {}}
  organization={
    Object {
      "fulfilled": true,
      "pending": false,
    }
  }
  quota={Object {}}
>
  <div
    className="spinner-fit-container"
  >
    <Spinner>
      <div
        className="ins-c-spinner"
        role="status"
      >
        <span
          className="pf-u-screen-reader"
        >
          Loading...
        </span>
      </div>
    </Spinner>
  </div>
  <div
    className="spinner-loading-text"
  >
    Loading node types...
  </div>
</MachineTypeSelection>
`;

exports[`<MachineTypeSelection /> when there was an error renders correctly 1`] = `
<MachineTypeSelection
  getMachineTypes={
    [MockFunction] {
      "calls": Array [
        Array [],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  input={
    Object {
      "onChange": [MockFunction] {
        "calls": Array [
          Array [
            "",
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      },
    }
  }
  isMultiAz={false}
  machineTypes={
    Object {
      "error": true,
      "errorMessage": "This is an error message",
      "fulfilled": false,
      "machineTypes": Array [],
      "pending": false,
    }
  }
  meta={Object {}}
  organization={
    Object {
      "fulfilled": true,
      "pending": false,
    }
  }
  quota={Object {}}
>
  <ErrorBox
    message="Error loading node types"
    response={
      Object {
        "error": true,
        "errorMessage": "This is an error message",
        "fulfilled": false,
        "machineTypes": Array [],
        "pending": false,
      }
    }
  >
    <Component
      className="error-box"
      isInline={true}
      title="Error loading node types"
      variant="danger"
    >
      <ComponentWithOuia
        component={[Function]}
        componentProps={
          Object {
            "children": Array [
              <span>
                This is an error message
              </span>,
              Array [],
              <br />,
              <span>
                Operation ID: N/A
              </span>,
            ],
            "className": "error-box",
            "isInline": true,
            "title": "Error loading node types",
            "variant": "danger",
          }
        }
        consumerContext={null}
      >
        <Alert
          className="error-box"
          isInline={true}
          ouiaContext={
            Object {
              "isOuia": false,
              "ouiaId": null,
            }
          }
          title="Error loading node types"
          variant="danger"
        >
          <div
            aria-label="Danger Alert"
            className="pf-c-alert pf-m-inline pf-m-danger error-box"
          >
            <AlertIcon
              variant="danger"
            >
              <div
                className="pf-c-alert__icon"
              >
                <ExclamationCircleIcon
                  color="currentColor"
                  noVerticalAlign={false}
                  size="sm"
                  title={null}
                >
                  <svg
                    aria-hidden={true}
                    aria-labelledby={null}
                    fill="currentColor"
                    height="1em"
                    role="img"
                    style={
                      Object {
                        "verticalAlign": "-0.125em",
                      }
                    }
                    viewBox="0 0 512 512"
                    width="1em"
                  >
                    <path
                      d="M504 256c0 136.997-111.043 248-248 248S8 392.997 8 256C8 119.083 119.043 8 256 8s248 111.083 248 248zm-248 50c-25.405 0-46 20.595-46 46s20.595 46 46 46 46-20.595 46-46-20.595-46-46-46zm-43.673-165.346l7.418 136c.347 6.364 5.609 11.346 11.982 11.346h48.546c6.373 0 11.635-4.982 11.982-11.346l7.418-136c.375-6.874-5.098-12.654-11.982-12.654h-63.383c-6.884 0-12.356 5.78-11.981 12.654z"
                      transform=""
                    />
                  </svg>
                </ExclamationCircleIcon>
              </div>
            </AlertIcon>
            <h4
              className="pf-c-alert__title"
            >
              <span
                className="pf-u-screen-reader"
              >
                Danger alert:
              </span>
              Error loading node types
            </h4>
            <div
              className="pf-c-alert__description"
            >
              <span>
                This is an error message
              </span>
              <br />
              <span>
                Operation ID: N/A
              </span>
            </div>
          </div>
        </Alert>
      </ComponentWithOuia>
    </Component>
  </ErrorBox>
</MachineTypeSelection>
`;
